README.md
# Data-Visualization
visualizations in Python

1. Word Cloud <br />
   easy and fun <br />
   wrap a text into a word could <br />
   
2. Stock Candlesticks <br />
   get data from web, pandas_datareader <br />
   draw candlestick charts, mpl_finance <br />
   
   
   


word could/news.txt
U.S. stocks fell sharply on Friday after a stronger-than-expected jobs report sent interest rates higher.

The Dow Jones industrial average dropped 665.75 points to close at 25,520.96, capping off the index's sixth-largest points decline ever. The 30-stock index also fell below 26,000. Friday also marked the first time since June 2016 that the Dow fell at least 500 points.

The S&P 500 fell 2.1 percent and finished at 2,762.13, with energy as the worst-performing sector. The Nasdaq composite plunged 1.96 percent to 7,240.95 as a decline in Apple and Alphabet offset a strong gain in Amazon shares.

The Dow posted its worst day since June 2016. The S&P 500 and Nasdaq had their biggest one-day fall since September 2016 and August 2017, respectively.

"The key for the market today is rising interest rates," said Mike Baele, managing director at U.S. Bank Wealth Management. "The old adage is: 'Bull markets don't die of old age, they are killed by higher interest rates.' That looms large."

The U.S. economy added 200,000 jobs in January, according to the Bureau of Labor Statistics. Economists polled by Reuters expected growth of 180,000. Wages, meanwhile, rose 2.9 percent on an annualized basis.

The report sent interest rates higher. The benchmark 10-year yield rose to 2.85 percent on the back of the report, hitting a four-year high. Investors have been jittery about the recent rise in interest rates, worrying they may be rising too fast.

On Friday, the 30-year yield rose its highest level since March.

"The reaction in the bond market is due to the rise in average hourly earnings," said James Ragan, director of individual investor group research at D.A. Davidson. "I think the market is now thinking of the possibility that the Fed could raise rates four times this year rather than three."

The Federal Reserve has forecast three rate hikes for 2018.

Bank stocks fell as the yield curve widened. The SPDR S&P Bank exchange-traded fund, which tracks bank stocks, dropped 1.2 percent. Banks typically benefit from higher interest rates.

This has been a volatile week for U.S. stocks. The Cboe Volatility index, widely considered the best fear gauge in the market, rose from 11.08 this week to 17.31.

The Dow, S&P 500 and Nasdaq snapped four-week winning streaks. The indexes also posted their worst weekly performance in two years.

"We've been expecting a pullback for a while, said Gene Goldman, head of research at Cetera Financial. "Yes, earnings are strong and the economy is doing well, but markets just don't go straight up."

Wall Street also looked to the release of key corporate earnings. Exxon Mobil reported weaker-than-expected earnings on Friday, sending its stock lower.

Tech giant Apple reported better-than-expected quarterly results. But the stock fell 4.4 percent after the company said it expects profit margins of 38 percent to 38.5 percent, tighter than the expected 38.9 percent. Apple also reported lighter-than-expected iPhone sales for its previous quarter.

Google-parent Alphabet also reported quarterly results, with earnings per share missing expectations. The company's stock dropped 5.3 percent on the back of the report.

Amazon, meanwhile, surged to an all-time high on the back of its earnings report. The e-commerce giant said its Amazon Web Services sales — a key component for the company — hit $5.11 billion. Analysts polled by FactSet expected AWS revenue of $4.97 billion.

About halfway through the earnings season, most companies have posted upside surprises. Of the S&P 500 companies that have reported as of Friday morning, 78 percent have beaten bottom-line expectations, while 80 percent have surpassed sales estimates, according to Thomson Reuters I/B/E/S.

Art Cashin, director of floor operations at UBS, said the release of the much-anticipated memo from House Intelligence Committee Chairman Devin Nunes on the Russia probe remained in the background as stocks sold off Friday.

"You're going into the weekend. That's inhibiting buyers more than creating sellers," Cashin said.



word could/wcgen.py

from os import path
from PIL import Image
import numpy as np
import matplotlib.pyplot as plt

from wordcloud import WordCloud, STOPWORDS

d = path.dirname(__file__)

# Read the whole text.
text = open(path.join(d, 'news.txt')).read()

# read the mask image
# taken from
# http://www.stencilry.org/stencils/movies/alice%20in%20wonderland/255fk.jpg
cloud_mask = np.array(Image.open(path.join(d, "cloud.png")))

stopwords = set(STOPWORDS)
#stopwords.add("said")

wc = WordCloud(background_color="white", max_words=150, mask=cloud_mask,
               stopwords=stopwords)
# generate word cloud
wc.generate(text)

# store to file
wc.to_file(path.join(d, "wccharm.png"))

# show
plt.imshow(wc, interpolation='bilinear')
plt.axis("off")
plt.figure()
plt.imshow(cloud_mask, cmap=plt.cm.gray, interpolation='bilinear')
plt.axis("off")
plt.show()


Stock candlesticks/candleVol.py
import pandas as pd
from datetime import datetime
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('whitegrid')

import matplotlib.dates as mdates
from mpl_finance import candlestick_ohlc


df = pd.read_csv('INTC2018.csv')

df['Date'] = df['Date'].apply(lambda x: datetime.strptime(x, '%Y-%m-%d'))

df['Date'] = df['Date'].map(mdates.date2num)
ohlc = df[['Date', 'Open', 'High', 'Low', 'Close']]

ma_day = [5, 10, 20, 50]

for ma in ma_day:
    col_name = "SMA%s" % (str(ma))
    df[col_name] = df['Adj Close'].rolling(window=ma, center=False).mean()


plt.figure(figsize=(10, 5))
top = plt.subplot2grid((4,4), (0, 0), rowspan=3, colspan=4)


candlestick_ohlc(top, ohlc.values, width=0.6, colorup='green', colordown='red')

top.xaxis.set_major_formatter(mdates.DateFormatter('%Y-%m-%d'))


top.plot(df['Date'], df['SMA5'], color='yellow', label='SMA5')
top.plot(df['Date'], df['SMA10'], color='blue', label='SMA10')
top.plot(df['Date'], df['SMA20'], color='pink', label='SMA20')
top.plot(df['Date'], df['SMA50'], color='brown', label='SMA50')

top.legend(loc='best')
top.set_title('INTC chart with volume')
#top.axes.get_xaxis().set_visible(False)


bottom = plt.subplot2grid((4,4), (3,0), colspan=4, rowspan=1)
bottom.bar(df.index, df['Volume'])
bottom.set_ylabel('Volume')


plt.tight_layout()
plt.show()







Stock candlesticks/candle.py
#!/usr/bin/env python3
# built on 09/08/2018
# pip install mpl_finance


import pandas as pd
from datetime import datetime
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('whitegrid')

import matplotlib.dates as mdates
from mpl_finance import candlestick_ohlc


df = pd.read_csv('AMD.csv')
print(df.info())
print(df.head())

###df['Date'] = pd.to_datetime(df['Date'])
df['Date'] = df['Date'].apply(lambda x: datetime.strptime(x, '%Y-%m-%d'))

df['Date'] = df['Date'].map(mdates.date2num)
ohlc = df[['Date', 'Open', 'High', 'Low', 'Close']]

df['ema5'] = df['Close'].ewm(span=5, adjust=False).mean()
df['ema10'] = df['Close'].ewm(span=10, adjust=False).mean()
df['ema20'] = df['Close'].ewm(span=20, adjust=False).mean()
df['ema50'] = df['Close'].ewm(span=50, adjust=False).mean()

fig, ax = plt.subplots(figsize=(12, 5))

candlestick_ohlc(ax, ohlc.values, width=0.6, colorup='green', colordown='red')
ax.xaxis.set_major_formatter(mdates.DateFormatter('%Y-%m-%d'))

ax.plot(df['Date'], df['ema5'], color='yellow', label='EMA5')
ax.plot(df['Date'], df['ema10'], color='blue', label='EMA10')
ax.plot(df['Date'], df['ema20'], color='pink', label='EMA20')
ax.plot(df['Date'], df['ema50'], color='brown', label='EMA50')

ax.legend(loc='best')



plt.savefig('AMD chart with EMA')
plt.show()

##SMA using rolling

#ma_day = [5, 10, 20, 50]
#for ma in ma_day:
#    col_name = "SMA%s" % (str(ma))
#    df[col_name] = AMD['Adj Close'].rolling(window=ma, center=False).mean()


###Monte Carlo simulation

# days = 100
# dt = 1/100
# mu = AMD.mean()['Adj Close']
# sigma = AMD['Adj Close'].std()
#
#
# def stock_monte_carlo(start_price, days, mu, sigma):
#     price = np.zeros(days)
#     price[0] = start_price
#
#     shock = np.zeros(days)
#     drift = np.zeros(days)
#
#     for x in range(1, days):
#         shock[x] = np.random.normal(loc=mu*dt, scale=sigma*np.sqrt(dt))
#
#         drift[x] = mu * dt
#
#         price[x] = price[x-1] + (price[x-1] * (drift[x] + shock[x]))
#
#     return price
#
# start_price = 11.49
#
# for run in range(100):
#     plt.plot(stock_monte_carlo(start_price, days, mu, sigma))
#
#
# runs = 1000
#
# simulations = np.zeros(runs)
#
#for run in range(runs):
#    simulations[run] = stock_monte_carlo(start_price,days,mu,sigma)[days-1]
#
# q = np.percentile(simulations,1)
# plt.hist(simulations,bins=200)
# plt.figtext(0.6,0.8,s="Start price: $%.2f" %start_price)
# plt.figtext(0.6,0.7,"Mean final price: $%.2f" % simulations.mean())
# plt.figtext(0.6,0.6,"VaR(0.99): $%.2f" % (start_price -q,))
# plt.figtext(0.15,0.6, "q(0.99): $%.2f" % q)
# plt.axvline(x=q, linewidth=4, color='r')
# plt.title(u"Final price distribution for Google Stock after %s days" %days, weight='bold')
# plt.show()

